src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+    if (lock == NULL)
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_002: [ If there are any failures, channel_internal_create_and_open shall fail and return NULL. ]*/
+        LogError("Failure in srw_lock_create(false, \"channel\")");
+    }
+    else
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_078: [ channel_create shall create a CHANNEL_INTERNAL object by calling THANDLE_MALLOC with channel_internal_dispose as dispose.]*/
+        THANDLE(CHANNEL_INTERNAL) channel_internal = THANDLE_MALLOC(CHANNEL_INTERNAL)(channel_internal_dispose);
+        if (channel_internal == NULL)
+        {
+            /*Codes_SRS_CHANNEL_INTERNAL_43_002: [ If there are any failures, channel_internal_create_and_open shall fail and return NULL. ]*/
+            LogError("Failure in THANDLE_MALLOC(CHANNEL_INTERNAL)(channel_internal_dispose)");
+        }
+        else
+        {
+            CHANNEL_INTERNAL* channel_internal_ptr = THANDLE_GET_T(CHANNEL_INTERNAL)(channel_internal);
+
+            channel_internal_ptr->lock = lock;
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_080: [ channel_create shall store given threadpool in the created CHANNEL_INTERNAL. ]*/
+            THANDLE_INITIALIZE(THREADPOOL)(&channel_internal_ptr->threadpool, threadpool);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_084: [ channel_create shall call DList_InitializeListHead. ]*/
+            DList_InitializeListHead(&channel_internal_ptr->op_list);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_086: [ channel_create shall succeed and return the created THANDLE(CHANNEL). ]*/
+            THANDLE_INITIALIZE_MOVE(CHANNEL_INTERNAL)(&result, &channel_internal);
+            goto all_ok;
+        }
+        srw_lock_destroy(lock);
+    }
+all_ok:
+    return result;
+}
+
+static void execute_callbacks(void* context)
+{
+    if (context == NULL)
+    {
+        LogError("Invalid arguments: void* context=%p", context);
+    }
+    else
+    {
+        CHANNEL_OP* channel_op = (CHANNEL_OP*)context;
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_145: [ execute_callbacks shall call the stored callback(s) with the result of the operation. ]*/
+        if (channel_op->pull_callback)
+        {
+            channel_op->pull_callback(channel_op->pull_context, channel_op->result, channel_op->data);
+        }
+        if (channel_op->push_callback)
+        {
+            channel_op->push_callback(channel_op->push_context, channel_op->result);
+        }
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_147: [ execute_callbacks shall perform cleanup of the operation. ]*/
+        THANDLE_ASSIGN(RC_PTR)(&channel_op->data, NULL);
+        THANDLE_ASSIGN(CHANNEL_INTERNAL)(&channel_op->channel_internal, NULL);
+
+        // copy to local reference to avoid cutting the branch you are sitting on
+        THANDLE(ASYNC_OP) temp = NULL;
+        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(&temp, &channel_op->async_op);
+        THANDLE_ASSIGN(ASYNC_OP)(&temp, NULL);
```

Comment:
I do not understand why these lines of code are needed.

They seem to do 

```
THANDLE_ASSIGN(ASYNC_OP)(&channel_op->async_op NULL);
```

Reply:
`channel_op` is the same as `async_op->context`.

`channel_op` also holds a `THANDLE(ASYNC_OP)` reference. So in a way, the async_op holds a reference to itself. 

If we did:

```c
THANDLE_ASSIGN(ASYNC_OP)(&channel_op->async_op NULL);
```

Then the memory for the async_op would get freed before the NULL assignment would happen. The NULL assignment would be an invalid access.